src/core/Akka/Actor/ReceiveActorHandlers.cs

```diff
@@ -0,0 +1,198 @@
+// -----------------------------------------------------------------------
+//  <copyright file="ReceiveActorHandlers.cs" company="Akka.NET Project">
+//      Copyright (C) 2009-2025 Lightbend Inc. <http://www.lightbend.com>
+//      Copyright (C) 2013-2025 .NET Foundation <https://github.com/akkadotnet/akka.net>
+//  </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Akka.Actor;
+#nullable enable
+internal sealed class ReceiveActorHandlers
+{
+    private bool _hadObjectHandlerWithNoPredicate;
+
+    public ReceiveActorHandlers()
+    {
+        TypedHandlers = new List<ITypeHandler>();
+        HandleAny = null;
+    }
+
+    private List<ITypeHandler> TypedHandlers { get; }
+
+    private Action<object>? HandleAny { get; set; }
+
+    private void CanAddMoreHandlers()
+    {
+        if (_hadObjectHandlerWithNoPredicate)
+        {
+            throw new InvalidOperationException("A handler for object with no predicate has already been added. No more handlers can be added as they would be ignored.");
+        }
+
+        if (HandleAny != null)
+        {
+            throw new InvalidOperationException("A handler that catches all messages has been added. No more handlers can be added as they would be ignored.");
+        }
+    }
+    
+    private static ITypeHandler CreateTypeHandler<T>(Predicate<T>? shouldHandlePredicate, Func<T, bool> handler)
+    {
+        if (shouldHandlePredicate == null)
+        {
+            return new TypeHandler<T>(handler);
+        }
+
+        return new PredicateHandler<T>(shouldHandlePredicate, handler);
+    }
+    
+    private static ITypeHandler CreateTypeHandler(Type t, Predicate<object>? shouldHandlePredicate, Func<object, bool> handler)
+    {
+        if (shouldHandlePredicate == null)
+        {
+            return new WeaklyTypedHandler(t, handler);
+        }
+
+        return new WeaklyTypedPredicateHandler(t, shouldHandlePredicate, handler);
+    }
+    
+    public void AddGenericReceiveHandler<T>(Predicate<T>? shouldHandlePredicate, Func<T, bool> handler)
+    {
+        CanAddMoreHandlers();
+        
+        TypedHandlers.Add(CreateTypeHandler(shouldHandlePredicate, handler));
+    }
+    
+
+    public void AddTypedReceiveHandler(Type messageType, Predicate<object>? shouldHandlePredicate, Func<object, bool> handler)
+    {
+        CanAddMoreHandlers();
+        
+        TypedHandlers.Add(CreateTypeHandler(messageType, shouldHandlePredicate, handler));
+
+        // If the message type is object, then we need to track that we have added a handler with no predicate.
+        if (messageType == typeof(object) && 
```

Comment:
In theory we could also do this from `AddGenericReceiveHandler<T>` but IMHO we're better off authoring an analyzer and flagging this issue at compilation - we have the ability to do that easily now through Akka.Analyzers.